syntax = "proto3";
package cao_commands;

import "cao_common.proto";
import "cao_script.proto";

enum StructureType {
    SPAWN = 0;
}

message PlaceStructureCommand
{
    cao_common.WorldPosition position = 1;
    cao_common.Uuid ownerId = 2;
    StructureType ty = 3;
}

message UpdateEntityScriptCommand
{
    cao_common.Uuid userId = 1;
    uint32 entityId = 2;
    cao_common.Uuid scriptId = 3;
}

message UpdateScriptCommand
{
    cao_common.Uuid userId = 1;
    cao_common.Uuid scriptId = 2;
    cao_script.CompilationUnit compilationUnit = 3;
}

message SetDefaultScriptCommand
{
    cao_common.Uuid userId = 1;
    cao_common.Uuid scriptId = 2;
}

message CommandResult
{
}

message TakeRoomCommand
{
    cao_common.Uuid userId = 1;
    cao_common.Axial roomId = 2;
}

message RegisterUserCommand
{
    cao_common.Uuid userId = 1;
    uint32 level = 2;
}

service Command
{
    rpc PlaceStructure(PlaceStructureCommand) returns (CommandResult) { }
    rpc UpdateEntityScript(UpdateEntityScriptCommand) returns (CommandResult) { }
    rpc UpdateScript(UpdateScriptCommand) returns (CommandResult) { }
    rpc SetDefaultScript(SetDefaultScriptCommand) returns (CommandResult) { }
    rpc TakeRoom(TakeRoomCommand) returns (CommandResult) { }
    rpc RegisterUser(RegisterUserCommand) returns (CommandResult) { }
}
